python/cudf/cudf/core/udf/templates.py

```diff
@@ -52,3 +57,22 @@ def _kernel(retval, size, input_col_0, offset_0, {extra_args}):
         ret_data_arr[i] = ret_masked.value
         ret_mask_arr[i] = ret_masked.valid
 """
+
+groupby_apply_kernel_template = """
+def _kernel(offset, out, index, {input_columns}, {extra_args}):
+    tid = cuda.threadIdx.x
+    block_id = cuda.blockIdx.x
+    tb_size = cuda.blockDim.x
+
+    recarray = cuda.local.array(1, dtype=dataframe_group_type)
+    dataframe_group = recarray[0]
+
+    if block_id < (len(offset) - 1):
+
+        size = offset[block_id+1] - offset[block_id]
+        arr_index = index[offset[block_id]:offset[block_id+1]]
+
+{group_initializers}
```

Comment:
So caller must ensure that `group_initializers` is appropriately indented.

Reply:
Yes. For now this follows the template built around `masked_initializer` above. This could be changed if we want but I'd rather change it everywhere.